File: src/hipo/SegmentHeader.cpp
//
// Copyright 2020, Jefferson Science Associates, LLC.
// Subject to the terms in the LICENSE file found in the top-level directory.
//
// EPSCI Group
// Thomas Jefferson National Accelerator Facility
// 12000, Jefferson Ave, Newport News, VA 23606
// (757)-269-7100


#include "SegmentHeader.h"


namespace evio {


    /**
     * Constructor.
     * @param tag the tag for the segment header.
     * @param dataType the data type for the content of the segment.
     */
    SegmentHeader::SegmentHeader(uint16_t tag, DataType const & dataType) :
                    BaseStructureHeader(tag, dataType) {}


    /**
     * Get the length of the structure's data in 32 bit ints (not counting the header words).
     * @return Get the length of the structure's data in 32 bit ints (not counting the header words).
     */
    uint32_t SegmentHeader::getDataLength() {return length;}


    /**
      * Get the length of the structure's header in ints. This includes the first header word itself
      * (which contains the length) and in the case of banks, it also includes the second header word.
      * @return Get the length of the structure's header in ints.
      */
    uint32_t SegmentHeader::getHeaderLength() {return 1;}


    /**
     * Write myself out as evio format data
     * into the given byte array in the specified byte order.
     *
	 * @param bArray array into which evio data is written (destination).
	 * @param order  byte order in which to write the data.
     * @return the number of bytes written, which for a SegmentHeader is 4.
     * @throws EvioException if destination array too small to hold data.
     */
    size_t SegmentHeader::write(uint8_t *dest, ByteOrder const & order) {

        auto word = (uint32_t) (tag << 24 | (uint8_t)((dataType.getValue() & 0x3f) |
                                (padding << 6)) << 16 | (length & 0xffff));
        Util::toBytes(word, order, dest);
        return 4;
    }


    /**
     * Write myself out a byte buffer.
     * This write is relative - i.e., it uses the current position of the buffer.
     *
     * @param byteBuffer the byteBuffer to write to.
     * @return the number of bytes written, which for a SegmentHeader is 4.
     */
    size_t SegmentHeader::write(std::shared_ptr<ByteBuffer> & byteBuffer) {
        return write(*(byteBuffer.get()));
    }


    /**
     * Write myself out a byte buffer.
     * This write is relative - i.e., it uses the current position of the buffer.
     *
     * @param byteBuffer the byteBuffer to write to.
     * @return the number of bytes written, which for a SegmentHeader is 4.
     */
    size_t SegmentHeader::write(ByteBuffer & byteBuffer) {

        auto word = (uint32_t) (tag << 24 | (uint8_t)((dataType.getValue() & 0x3f) |
                                (padding << 6)) << 16 | (length & 0xffff));
        byteBuffer.putInt(word);
        return 4;
    }


    /**
     * Obtain a string representation of the segment header.
     * @return a string representation of the segment header.
     */
    std::string SegmentHeader::toString() {
        std::stringstream ss;

        ss << "segment length: " << length << std::endl;
        ss << "     data type: " << getDataTypeName() << std::endl;
        ss << "           tag: " << tag << std::endl;
        ss << "       padding: " << +padding << std::endl;

        return ss.str();
    }

}

File: java/org/jlab/coda/jevio/SegmentHeader.java
package org.jlab.coda.jevio;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * This the the header for an evio segment structure (<code>EvioSegment</code>). It does not contain the raw data, just the
 * header. 
 * 
 * @author heddle
 * 
 */
public final class SegmentHeader extends BaseStructureHeader {
	
	
	/**
	 * Null constructor.
	 */
	public SegmentHeader() {
	}

	/**
	 * Constructor.
	 * @param tag the tag for the segment header.
	 * @param dataType the data type for the content of the segment.
	 */
	public SegmentHeader(int tag, DataType dataType) {
		super(tag, dataType);
	}

    /**
     * {@inheritDoc}
     */
    public int getHeaderLength() {return 1;}

    /**
     * {@inheritDoc}
     */
    protected void toArray(byte[] bArray, int offset, ByteOrder order) {
        try {
			int word = (tag << 24 | (byte)((dataType.getValue() & 0x3f) |
				       (padding << 6)) << 16 | (length & 0xffff));
			ByteDataTransformer.toBytes(word, order, bArray, offset);
        }
        catch (EvioException e) {e.printStackTrace();}
    }

	/**
	 * Obtain a string representation of the segment header.
	 * @return a string representation of the segment header.
	 */
	@Override
	public String toString() {
		return String.format("segment length: %d\ndata type:      %s\ntag:            %d\npadding:        %d\n",
				length, getDataTypeName(), tag, padding);
	}
	
	/**
	 * Write myself out a byte buffer. This write is relative--i.e., it uses the current position of the buffer.
	 * 
	 * @param byteBuffer the byteBuffer to write to.
	 * @return the number of bytes written, which for a SegmentHeader is 4.
	 */
	public int write(ByteBuffer byteBuffer) {
		int word = (tag << 24 | (byte)((dataType.getValue() & 0x3f) |
				   (padding << 6)) << 16 | (length & 0xffff));
		byteBuffer.putInt(word);
		return 4;
	}


}